### PR TITLE
Fix the "Use this Display" button so that secondary displays can be disabled (#131)

### LINKED ISSUE #131: "Use this Display" switch doesn't work
I just installed Juno and I noticed that disabling my ThinkPad X230's internal display doesn't work. It just stays on.

Steps to reproduce:
Boot into Juno install media or install it and then:
1. Connect an external monitor to your laptop
2. System Settings -> Display
3. Flip the "Use this Display" switch on your internal display
4. Apply the changes
5. See how nothing has changed at all

### PR DESCRIPTION
**Pull request by shavitmichael**
Fixes #131
Monitors are disabled by excluding them from any logical monitors during the ApplyMonitorsConfig call.

### CODE REVIEW

**Comment by jeremypw** [comment]
Thanks for the contribution! There is some lint - please check your changes pass testing with io.elementary.vala-lint.  In particular there should be a space before an opening bracket.  Don't worry about any pre-existing lint, we just do not want to add more here.

You have introduced a property `is_active` with an automatic setter but also introduced an explicit function `set_active` which sets this property.  Do you need both ways of setting this property? It seems prone to conflict although I have not looked at it in detail yet.  

Please confirm that all changes made are essential to fixing the linked issue.

**Comment by jeremypw** [comment]
Another issue:

* If a suspend and restore intervenes, re-enabling the internal monitor does not work properly. 
- The confirmatory dialog is not always visible the first time (have to confirm with keyboard)
- After re-enabling, the monitor does not actually switch on the first time.

* If two suspends intervene with usage without the internal monitors used between suspends, I got to a state where I could not get the internal monitor to switch back on after many attempts even with logouts and reboots. Eventually I noticed that the properties menu of the internal monitor showed blank in the refresh-rate combo.  When I selected a refresh rate then the monitor did switch on.

Nevertheless the login screen always appeared on the internal monitor OK.

**Comment by jeremypw** [comment]
This does not seem to work for one of my external monitors (connected to HDMI), but the other monitor (connected to VGA), switched on and off without problem.  Not sure if this is a hardware or software issue at the moment.

This PR is probably exposing issues in other parts of the code unfortunately and we can only allow those operations where all serious knock-on effects have been addressed. :-(

**Comment by shavitmichael** [comment]
Thanks for the review!

> Thanks for the contribution! There is some lint - please check your changes pass testing with io.elementary.vala-lint. In particular there should be a space before an opening bracket. Don't worry about any pre-existing lint, we just do not want to add more here.

Done :).

> You have introduced a property is_active with an automatic setter but also introduced an explicit function set_active which sets this property. Do you need both ways of setting this property? It seems prone to conflict although I have not looked at it in detail yet.

Is there a canonical way of providing a custom setter implementation in valac? Would I have to explicitly define a member variable for it?
I've made the default setter private for now, pending a better solution.

> * If a suspend and restore intervenes, re-enabling the internal monitor does not work properly.
> 
> * After re-enabling, the monitor does not actually switch on the first time.
> 
> * If two suspends intervene with usage without the internal monitors used between suspends, I got to a state where I could not get the internal monitor to switch back on after many attempts even with logouts and reboots. Eventually I noticed that the properties menu of the internal monitor showed blank in the refresh-rate combo.  When I selected a refresh rate then the monitor did switch on.

I wasn't able to reproduce these exactly, but I did run into similar issues where enabling a monitor would no longer work. I found that we were failing the set_configuration call because of incorrect ar

> By closing and reopening the laptop lid It was possible to get to a state where the monitor appeared disabled in the plug but was actually switched on.

Latest commit now redraws the window when a config change is detected.

**Comment by jeremypw** [comment]
>Is there a canonical way of providing a custom setter implementation in valac? Would I have to explicitly define a member variable for it?

Yes, the usual way of doing it is to have a private member holding the current value of the property e.g.:

```
private bool _is_active = true;
public bool is_active {
    get {
        return _is_active;
    }

    set {
        _is_active = is_mirror || value; // Aways active if is mirrored
    }
}
```

However, you cannot return a result this way.  If you really need to do this it would be more consistent to have a public getter function as well and just use a private member to hold the value. Otherwise you have to access the value with one syntax and change it with a different syntax.

File: src/Widgets/DisplayWidget.vala [hunk 1]
```diff
@@ -205,7 +205,9 @@ public class Display.DisplayWidget : Gtk.EventBox {
         destroy.connect (() => display_window.destroy ());
 
         use_switch.notify["active"].connect (() => {
-            //output_info.set_active (use_switch.active);
+            if (!virtual_monitor.set_active (use_switch.active)) {
```
Comment (jeremypw) [inline src/Widgets/DisplayWidget.vala]: You should bind the `use_switch.active` "sensitive" property to `virtual_monitor.is_mirrored` with an `INVERT_BOOLEAN` binding flag.  That way the user would not be able to turn off a mirrored monitor and no boolean setter is needed.

File: src/Widgets/DisplayWidget.vala [hunk 2]
```diff
@@ -205,7 +205,9 @@ public class Display.DisplayWidget : Gtk.EventBox {
...
+                return;
+            };
             resolution_combobox.sensitive = use_switch.active;
             rotation_combobox.sensitive = use_switch.active;
             refresh_combobox.sensitive = use_switch.active;
```
Comment (jeremypw) [inline src/Widgets/DisplayWidget.vala]: It would be better to bind these properties to `virtual_monitor.is_active`.

**Comment by jeremypw** [comment]
It really needs a thorough review of how properties of the various widgets are synchronised with virtual monitor properties (this matters more now you can deactivate monitors). Use `bind_property ()` where possible. The comboboxes default active iters should be set as soon as the models are populated so that they cannot ever be `-1`.

File: src/Widgets/DisplayWidget.vala (hunk 1 above)
Comment (shavitmichael) [reply src/Widgets/DisplayWidget.vala]: That worked great, thanks!

File: src/Widgets/DisplayWidget.vala (hunk 2 above)
Comment (shavitmichael) [reply src/Widgets/DisplayWidget.vala]: hmm, I tried replacing these in the same fashion as the use_switch.sensitive property but I couldn't get it to work, e.g:
` virtual_monitor.bind_property ("is_active", rotation_combobox, "sensitive");
`
When I tested this, the boxes wouldn't toggle visibility with the active button; they stayed stuck on their initial state. 

To be honest I've already spent wayy more time on this PR than I was expecting, so I'm tempted to punt this issue aside. What do you think?
In the mean time I've changed those lines to at least be set from virtual_monitor.is_active.

Comment (tintou) [reply src/Widgets/DisplayWidget.vala]: You have to replace underscores with hyphen so:
```vala
virtual_monitor.bind_property ("is-active", rotation_combobox, "sensitive");
```

**Comment by jeremypw** [comment]
@shavitmichael  Please do not despair! We appreciate you efforts.  This should be close to merging now, but if you really have no more time to work on it an elementary dev could adopt it.

**Comment by jeremypw** [comment]
@shavitmichael I am working on a PR for this branch which will (hopefully) finish it.

**Comment by jeremypw** [comment]
There are so many serious bugs in the master code it makes it hard to focus on fixing just this PR and determining whether isuues found are caused by this PR or not!  I'll submit some other PRs as well I think.

**Comment by jeremypw** [comment]
I'll have very limited time until the weekend though :-(

**Comment by shavitmichael** [comment]
Thanks I definitely appreciate the help!
I also found some time fix a minor issue with the "set primary" button on disabled monitors; we're definitely getting closer :) .

**Comment by jeremypw** [comment]
I started a branch here: https://github.com/elementary/switchboard-plug-display/tree/various-fixes, but I haven't had time to delint, test fully and separate out those fixes directly related to this issue yet.  Please take from it whatever you like (if you need to). I will not have time to review your latest changes until Saturday.

File: src/Objects/MonitorManager.vala [hunk 3]
```diff
@@ -20,13 +20,13 @@
  */
 
 public class Display.MonitorManager : GLib.Object {
-    public Gee.LinkedList<Display.VirtualMonitor> virtual_monitors { get; construct; }
-    public Gee.LinkedList<Display.Monitor> monitors { get; construct; }
-
+    public Gee.LinkedList<Display.VirtualMonitor> virtual_monitors { get; set; }
+    public Gee.LinkedList<Display.Monitor> monitors { get; set; }
```
Comment (jeremypw) [inline src/Objects/MonitorManager.vala]: Why can these not be `construct` properties?  You can still add to the lists after constructing. You do not need to set new list objects though.

File: src/Objects/MonitorManager.vala [hunk 4]
```diff
@@ -66,12 +66,22 @@ public class Display.MonitorManager : GLib.Object {
...
-            iface.monitors_changed.connect (get_monitor_config);
+            iface.monitors_changed.connect (on_monitors_changed);
         } catch (Error e) {
             critical (e.message);
         }
     }
 
+    public void rescan_monitors () {
+        monitors.clear ();
+        virtual_monitors.clear ();
```
Comment (jeremypw) [inline src/Objects/MonitorManager.vala]: Shouldn't you notify this change in virtual monitor size?

File: src/Objects/MonitorManager.vala [hunk 5]
```diff
@@ -66,12 +66,22 @@ public class Display.MonitorManager : GLib.Object {
         virtual_monitors = new Gee.LinkedList<Display.VirtualMonitor> ();
         try {
             iface = Bus.get_proxy_sync (BusType.SESSION, "org.gnome.Mutter.DisplayConfig", "/org/gnome/Mutter/DisplayConfig");
-            iface.monitors_changed.connect (get_monitor_config);
+            iface.monitors_changed.connect (on_monitors_changed);
```
Comment (jeremypw) [inline src/Objects/MonitorManager.vala]: Better name `on_iface_monitors_changed` to distinguish from local signal with a similar name.

File: src/Objects/MonitorManager.vala [hunk 6]
```diff
@@ -206,13 +216,19 @@ public class Display.MonitorManager : GLib.Object {
         }
         if (virtual_monitor_number != monitor_number) {
             // Create Inactive VirtualMonitors for each disabled monitor.
-            virtual_monitors.add_all(create_missing_virtual_monitors(false, max_scale));
+            virtual_monitors.add_all (create_missing_virtual_monitors (false, max_scale));
```
Comment (jeremypw) [inline src/Objects/MonitorManager.vala]: Shouldn't this change in size be notified?

File: src/Objects/MonitorManager.vala [hunk 7]
```diff
@@ -174,7 +184,7 @@ public class Display.MonitorManager : GLib.Object {
                 }
             }
         }
-
+        double max_scale = Utils.get_min_compatible_scale (monitors);
```
Comment (jeremypw) [inline src/Objects/MonitorManager.vala]: Don't you need to get the *maximum* compatible scale? You will need a new Utils function I think,

File: src/Objects/MonitorManager.vala [hunk 8]
```diff
@@ -188,6 +198,7 @@ public class Display.MonitorManager : GLib.Object {
             virtual_monitor.current_x = mutter_logical_monitor.x;
             virtual_monitor.current_y = mutter_logical_monitor.y;
             virtual_monitor.scale = mutter_logical_monitor.scale;
+            max_scale = virtual_monitor.scale;
```
Comment (jeremypw) [inline src/Objects/MonitorManager.vala]: Did you mean this? Or `max_scale = double.min (max_scale, virtual_monitor.scale);`

File: src/Views/DisplaysView.vala [hunk 9]
```diff
@@ -110,7 +110,8 @@ public class Display.DisplaysView : Gtk.Grid {
                 mirror_grid.sensitive = monitor_manager.monitors.size > 1;
```
Comment (jeremypw) [inline src/Views/DisplaysView.vala]: I think the mirror switch should be bound to the monitor_manager "is_mirrored" property not the number of monitors?  Not strictly part of this PR though.

**Comment by jeremypw** [comment]
I think I was confusing "scale" and "resolution (mode)" so I have removed a couple of comments.

File: src/Widgets/DisplayWidget.vala [hunk 10]
```diff
@@ -203,12 +205,12 @@ public class Display.DisplayWidget : Gtk.EventBox {
         display_window.attached_to = this;
 
         destroy.connect (() => display_window.destroy ());
-
+        virtual_monitor.bind_property ("is_mirror", use_switch, "sensitive", GLib.BindingFlags.INVERT_BOOLEAN);
```
[the remark this answers is not in the document]
Comment (shavitmichael) [reply src/Widgets/DisplayWidget.vala]: This was resolved in https://github.com/elementary/switchboard-plug-display/pull/197/commits/1152b5fa4983bd2060b52c8e6838b768e4090feb

File: src/Widgets/DisplayWidget.vala [hunk 11]
```diff
@@ -206,9 +206,9 @@ public class Display.DisplayWidget : Gtk.EventBox {
         virtual_monitor.bind_property ("is_mirror", use_switch, "sensitive", GLib.BindingFlags.INVERT_BOOLEAN);
         use_switch.notify["active"].connect (() => {
             virtual_monitor.is_active = use_switch.active;
-            resolution_combobox.sensitive = use_switch.active;
-            rotation_combobox.sensitive = use_switch.active;
-            refresh_combobox.sensitive = use_switch.active;
+            resolution_combobox.sensitive = virtual_monitor.is_active;
+            rotation_combobox.sensitive = virtual_monitor.is_active;
+            refresh_combobox.sensitive = virtual_monitor.is_active;
```
[the remark this answers is not in the document]
Comment (shavitmichael) [reply src/Widgets/DisplayWidget.vala]: This was resolved in https://github.com/elementary/switchboard-plug-display/pull/197/commits/1152b5fa4983bd2060b52c8e6838b768e4090feb

File: src/Objects/MonitorManager.vala (hunk 3 above)
Comment (shavitmichael) [reply src/Objects/MonitorManager.vala]: Ah, that might have been a leftover from a previous attempt. Removed.

File: src/Objects/MonitorManager.vala (hunk 5 above)
Comment (shavitmichael) [reply src/Objects/MonitorManager.vala]: Done.

File: src/Objects/MonitorManager.vala (hunk 4 above)
Comment (shavitmichael) [reply src/Objects/MonitorManager.vala]: Hmm, I'm not sure what you mean?

File: src/Objects/MonitorManager.vala (hunk 6 above)
Comment (shavitmichael) [reply src/Objects/MonitorManager.vala]: I'm also not sure what you mean here either.
These are newly created VirtualMonitor instances, so there aren't corresponding DisplayWidget to notify yet?

**Comment by shavitmichael** [comment]
Answered and resolved some of the inline comments in the latest commit. Some questions are still open however.
(Thanks again for the review and the help on the fixes! I've had a busy week and haven't had much time to work on this)

File: src/Objects/MonitorManager.vala (hunk 4 above)
Comment (jeremypw) [reply src/Objects/MonitorManager.vala]: I meant calling `notify["virtual-monitor-number"].  However, a better solution might be for the connections to be made directly to `virtual_monitors.size`  i.e. `monitor_anager.virtual_monitors.notify["size"].connect (() => ....` then no explicit notification would be needed.  However I would need to test this.

File: src/Objects/MonitorManager.vala (hunk 6 above)
Comment (jeremypw) [reply src/Objects/MonitorManager.vala]: OK, but see previous comment about connecting differently.

**Comment by jeremypw** [comment]
The main problem now is the messing up of the auto-layout when one or more of the monitors is disabled.  I find that the two remaining monitors appear vertically offset (and cannot be made to appear aligned) although moving the the mouse pointer horizontally from one to the other shows that they are in fact aligned.

**Comment by praneetloke** [comment]
It doesn't look like there has been much progress on this for sometime now. What is needed here to get this through? I would love to see this get merged myself (and so will others affected by this issue). Trying to see if I can help in some way.

**Comment by jeremypw** [comment]
There was an outstanding issue when I last looked at this but the author has not resolved it yet.  I'll have another look now that the latest master has been merged.  If you are able to build and install this branch and test it with two or more monitors and report your findings that would help.

**Comment by praneetloke** [comment]
>  If you are able to build and install this branch and test it with two or more monitors and report your findings that would help.

I'll try to do this. I haven't explored the repo yet, but are there instructions somewhere on how to test this on my own installation of the OS? I haven't done this sort of thing before, so I am a total newbie at this.

Also I would only be able to test this by turning off by laptop's screen, and have display on the external monitor; basically just one screen output.

**Comment by jeremypw** [comment]
There are basic instructions in the README.md on this page: https://github.com/elementary/switchboard-plug-display but maybe not enough for a "newbie".  More extensive information on installing the required tools are here: https://elementary.io/docs/code/getting-started#getting-started  particularly the sections on the Basic Setup and Build System.  Don't worry if it is too much - hopefully this will be progressed soon.

**Comment by praneetloke** [comment]
Alright. I managed to figure out how to build this locally and test it. It wasn't too bad after all.

Here's what I found:

#### Toggling the Use this Display switch causes a weird transition. Not sure if this is intended to be this way?

Screen record from 2020-02-11 21 11 16

#### Switching off the built-in display and only having the external monitor on

> Notice how the disabled display has moved to the right in the Display widget

Screenshot from 2020-02-11 21-14-20

The one nice thing I noticed is that the confirmation dialog now actually works. Previously, clicking the **Restore previous configuration** button did nothing for me, but now it correctly reverts to the previous configuration as I would expect it to.

**Comment by jeremypw** [comment]
@praneetloke Congrats on joining the ranks of source code builders and testers!

@shavitmichael Do you have the time to have a look at this PR again?  If not I'll try and progress it.

**Comment by shavitmichael** [comment]
Sorry for the lack of progress here, but I no longer have the time to work on this.

I've lost quite a bit of context, but IIRC this fix surfaced a bunch of underlying issues in the existing codebase. IMO, as long as the user can't get in a bad state because of this PR however, it could be worth merging it as it is. Sure the interface might be a little buggy, but it's probably better than a disable button that does nothing...

File: src/Objects/MonitorManager.vala (hunk 4 above)
Comment (praneetloke) [reply src/Objects/MonitorManager.vala]: I looked at resolving this review comment myself, and I actually don't see anything subscribing to virtual monitors' size. Maybe I am misunderstanding something?

**Comment by jeremypw** [comment]
I have pushed a PR (#228) based on this one that seems to solve the layout problems by displaying inactive displays separately as a button.

**Comment by shavitmichael** [comment]
That's awesome, thanks a lot for picking this bug back up Jeremy!

I'll close this PR and hope that yours moves moves forward :) .

**Pull request closed without merging**